utils/retrieval_evaluator.py

```diff
@@ -0,0 +1,627 @@
+import numpy as np
+from pathlib import Path
+import pandas as pd
+from matplotlib.gridspec import GridSpec
+from matplotlib.patches import Patch
+from scipy import stats
+import time
+import logging
+from fuzzywuzzy import fuzz
+from collections import defaultdict
+from typing import List, Dict, Any, Union
+import argparse
+from datetime import datetime
+import shutil
+from goodai.ltm.mem.base import RetrievedMemory
+
+from utils.constants import RETRIEVAL_REFERENCE_DIR
+from utils.util_math import get_dimensions
+
+
+class FileManager:
+    def __init__(self, base_directory: Path):
+        self.base_directory = Path(base_directory)
+        self.ensure_directory_exists(self.base_directory)
+        self.comparison_data_file = self.base_directory / "comparison_data.json"
+
+    def ensure_directory_exists(self, path):
+        path.mkdir(parents=True, exist_ok=True)
+
+    def generate_filename(self, prefix="comparison_data"):
+        timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
+        return f"{prefix}_{timestamp}.json"
+
+    def clear_comparison_data(self, output_dir: Path, timestamp: int):
+        if self.comparison_data_file.exists():
+            new_filename = f"old_comparison_data_{datetime.fromtimestamp(timestamp).strftime('%Y%m%d_%H%M%S')}.json"
+            new_file_path = output_dir / new_filename
+
+            shutil.move(str(self.comparison_data_file), str(new_file_path))
+
+            print(f"Previous comparison data moved to {new_file_path}")
+
+        with open(self.comparison_data_file, 'w') as f:
+            json.dump([], f)
+
+        print(f"Created new empty comparison data file at {self.comparison_data_file}")
+
+    def write_data(self, data):
+        try:
+            with open(self.comparison_data_file, 'w') as f:
+                json.dump(data, f, indent=2)
+            print(f"Data written to {self.comparison_data_file}")
+        except PermissionError as e:
+            print(f"Permission error when writing to {self.comparison_data_file}. Error: {e}")
+            fallback_path = Path.home() / "comparison_data_fallback.json"
+            with open(fallback_path, 'w') as f:
+                json.dump(data, f, indent=2)
+            print(f"Data written to fallback location: {fallback_path}")
+
+    def read_data(self):
+        if not self.comparison_data_file.exists():
+            print(f"Comparison data file not found at {self.comparison_data_file}")
+            return []
+        with open(self.comparison_data_file, 'r') as f:
+            return json.load(f)
+
+class RetrievalEvaluator:
+    def __init__(self):
+        self.project_root = Path(__file__).parent.parent / "data" / "retrieval_evaluator"
+        self.dev_bench_reference_data_path = RETRIEVAL_REFERENCE_DIR
+        self.comparison_data_path = self.project_root / "comparison_data"
+        self.logs_path = self.project_root / "logs"
+        self.evaluation_outputs_path = self.project_root / "evaluation_outputs"
```

Comment:
I think it would be very handy to be able to do something like `python evaluate_retrieval.py "Run name" "Agent name"` and have the code grab the full log and all, instead of having to go through a file.

Reply:
This would only work if all the memory retrievals were saved by default. While we are able to generate the reference data from the test defs and he master log for a run, we only get the retrieval stats if the agent gives them to us.

Since we treat the agents as a black box, there is no guarantee that this can work, and I don't want to institute boilerplate for them.

Reply:
I see. I oversimplified it in my head.